# 통계 Day 2과제 

## 필수 라이브러리 추가 및 폰트 설정

In [1]:
import pandas as pd
import numpy as np
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats
import scipy.stats

%matplotlib inline
matplotlib.rc('font', family='Malgun Gothic')
plt.rc('font', family = 'Malgun Gothic')

## 실습 1 p43

### 대한민국 남성의 20~30대 키를 측정한 결과, 평균 175cm 표준편차 5cm이다
 - 나의 키가 180cm일 경우에 나는 몇 % 범위일까?

In [2]:
mu = 175 #키 평균 (cm)
sigma = 5 #표준편차 (cm)

x = 180 #알고 싶은 값 (cm)

prob = stats.norm.cdf(x,mu,sigma) #cumulative distribution function 누적 분포 함수
print("{}cm 는 대한민국 남성 상위 {:.2f}%입니다.".format(x,(1-prob)*100))

180cm 는 대한민국 남성 상위 15.87%입니다.


180cm는 평균 175cm에서 5cm 떨어져 있고 여기서 표준편차가 5cm이므로 표준편차의 1 배의 위치에 위치한다. 따라서 180cm 사람은 상위 15.87%에 위치한다고 할 수 있다. 

## 실습 2 p45

### 어떤 자동차 1리터당 주행거리가 평균 12km, 표준편차 3km 인 정규 분포를 이룬다.
 - 1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 얼마인가요?
 - 1리터를 가지고 15km 이상을 달릴 확률은 얼마인가요?

In [3]:
mu = 12 #리터당 평균 주행거리 (km)
sigma = 3 #표준편차 (km)
x_min = 9
x_max = 15


prob1 = stats.norm.cdf(x_max,mu,sigma) - stats.norm.cdf(x_min,mu,sigma)
print("1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 {:.2f}% 입니다.".format(prob1*100))

prob2 = 1 - stats.norm.cdf(x_max,mu,sigma)
print("1리터를 가지고 15km 이상을 달릴 확률은 {:.2f}% 입니다.".format(prob2*100))

1리터를 가지고 9km 이상 15km 이하를 달릴 확률은 68.27% 입니다.
1리터를 가지고 15km 이상을 달릴 확률은 15.87% 입니다.


9km<x < 15km 의 경우 15km이하를 달릴 확률에서 9km이하를 달릴 확률을 빼주어 계산하였다.
15km이상을 달릴 확률은 전체 확률 1에서 15km이하를 달릴 확률을 빼주어 계산하였다.

## 실습 3 p54

### 확률변수t는 자유도가 100인 t-분포를 따른다.
 - t 값이 2.0일 때, P(T<t)을 계산하시오

In [4]:
t = 2.0 #확률변수 t
df = 100 #자유도

#누적 확률 값 계산
prob = stats.t.cdf(t,df) #cumulative distribution function 누적 분포 함수
print("P(T<=F): {:.2f}".format(prob))

P(T<=F): 0.98


## 실습 4 p60

### 확률변수 $\chi^2$ 은 자유도가 8인 $\chi^2$ 분포를 따른다 
 - $\chi^2$ 값이 3.0일 때, P(X<=$\chi^2$)을 계산하시오

In [5]:
chisq = 3.0 #카이스퀘어
df = 8 #자유도
prob = stats.chi2.cdf(chisq,df) #누적 분포 함수
print("P(X<={}) : {:.5f}".format(chisq,prob))

P(X<=3.0) : 0.06564


## 실습 5 p64

### 확률변수 F는 각각 자유도가 10, 100 인 F-분포를 따른다.
 - F값이 1.5 일 때, P(X<=F)을 계산하시오.

In [6]:
f = 1.5 #확률변수
dfnum = 10 #F(10,100)-분포
dfden = 100

prob = stats.f.cdf(x=f, dfn = dfnum, dfd = dfden) #누적 분포 함수
print("P(X<=F): {:.2f}".format(prob))

P(X<=F): 0.85


## 실습 6 p68

### 어떤 제품의 수명시간 x가 형상모수 5.2, 척도 모수 2,200 인 와이블 분포를 따른다고 할 때,
- 이 제품이 적어도 2,500 시간 이상 작동할 확률을 구하라

In [7]:
x = 2500 #시간
alpha = 5.2 #형상모수
beta = 2200 #척도모수

#누적확률 값 계산
prob = stats.weibull_min.cdf(x, alpha, scale = beta)
print("P(X<=x): {:.3f}".format(1-prob))

P(X<=x): 0.143


## 실습 7 p75

### 도장공정에서 광택도 불량이 30% 나 된다고 한다.
 - 5대의 차량을 임의로 선택했을 때 불량대수가 각각 0,1,2,3,4,5 대가 나올 확률은 얼마인가요?

In [8]:
n = 5
for i in range(n+1):
    #이항 분포 Probability Mass Function
    prob = stats.binom.pmf(k = i, n=n, p = 0.3)
    print("P(X={}) = {:.5f}".format(i,prob))

P(X=0) = 0.16807
P(X=1) = 0.36015
P(X=2) = 0.30870
P(X=3) = 0.13230
P(X=4) = 0.02835
P(X=5) = 0.00243


## 실습 8 p78

### 한 상점에는 평균 1시간에 3명씩 포아송분포로 손님이 도착한다.
 - 이때 10시에서 11시 사이에 다음이 발생할 확률을 구하시오
     - 2명의 손님이 도착
     - 1명 이하의 손님이 도착
     - 3명 이상의 손님이 도착

In [9]:
mu = 3 #평균 한시간에 3명 손님
prob1 = stats.poisson.pmf(2,mu) #1시간에 2명의 손님이 올 확률
prob2 = stats.poisson.cdf(1,mu) #1시간에 1명 이하의 손님이 올 확률 (누적 분포 함수)
prob3 = 1 - stats.poisson.cdf(2,mu) # 2명 이하의 손님이 올때를 구하여 1에서 뺀다

print("10시에서 11시 사이에 2명의 손님이 도착할 확률 : {:.2f}%".format(prob1*100))
print("10시에서 11시 사이에 1명 이하의 손님이 도착할 확률 : {:.2f}%".format(prob2*100))
print("10시에서 11시 사이에 3명 이상의 손님이 도착할 확률 : {:.2f}%".format(prob3*100))

10시에서 11시 사이에 2명의 손님이 도착할 확률 : 22.40%
10시에서 11시 사이에 1명 이하의 손님이 도착할 확률 : 19.91%
10시에서 11시 사이에 3명 이상의 손님이 도착할 확률 : 57.68%


## 실습 9 p96

### 실습파일 2sample_t_test.csv을 불러온다.
 - 변수, Mat_A 와 Mat_B에 대해 각각 정규성을 검정합니다.

In [10]:
from scipy.stats import shapiro

ds_Mat_AB_data = pd.read_csv('2sample_t_test.csv',engine = "python")
ds_Mat_AB_data.head()

data1 = ds_Mat_AB_data['Mat_A'].dropna()
shapiro(data1)

statistic, p = stats.shapiro(data1)
print("Shapiro-Wilk Test: statistic={:.4f}, p-value = {:.4f}".format(statistic,p))

Shapiro-Wilk Test: statistic=0.9063, p-value = 0.2563


불러온 데이터에서 NaN값을 제거해주기위해 dropna메소드를 사용하여 제거하였다.
신뢰수준 95%일때, 정규성 검정결과 p-value값이 0.05보다 크므로 정규로 판단 할 수 있다.

In [11]:
from scipy.stats import shapiro

ds_Mat_AB_data = pd.read_csv('2sample_t_test.csv',engine = "python")
ds_Mat_AB_data.head()

data2 = ds_Mat_AB_data['Mat_B'].dropna()
shapiro(data2)

statistic, p = stats.shapiro(data2)
print("Shapiro-Wilk Test: statistic={}, p-value = {:.4f}".format(statistic,p))


Shapiro-Wilk Test: statistic=0.9729753732681274, p-value = 0.9170


불러온 데이터에서 NaN값을 제거해주기위해 dropna메소드를 사용하여 제거하였다.
신뢰수준 95%일때, 정규성 검정결과 p-value값이 0.05보다 크므로 정규로 판단 할 수 있다.

## 실습 10 p99

### 새로운 품종의 감자 20 Box를 랜덤추출하여 무게를 측정할 결과가 다음과 같음
 - 단, 모표준편차는 3.8로 알려져 있음
 - 이 자료로부터 신품종 배추의 평균 무게에 대한 95% 신뢰구간을 구하시오.


In [12]:
df = pd.DataFrame({"sample" : [18, 18, 20, 21, 20, 23, 19, 18, 17, 21,
                               22, 20, 20, 21, 20, 19, 19, 18, 17, 19]})
df.head()
#정규분포 구간 추정 함수 실행
lower,upper = stats.norm.interval(0.95,loc = np.mean(df), scale = 3.8/np.sqrt(20))
print("신뢰구간: ({}, {})".format(lower.round(2),upper.round(2)))

신뢰구간: ([17.83], [21.17])


## 실습 11 p102

### 제품개발자는 새로 개발된 강종의 평균 인장강도를 70 kg/mm^2 으로 기대하고 있음.
- 이를 검증하기 위해 시험 생산되는 강종 중 임의로 20개의 코일을 대상으로 랜덤하게 시편을 채취하였음.
- 이 자료로 부터 표본 평균은 71.50kg/mm^2, 표본표준편차는 1.933 kg/mm^2 임을 알았음.
- 이 모평균 값의 95% 신뢰 구간을 추정하시오

In [13]:
df = pd.DataFrame({ 'sample':[73, 71, 74, 69, 70, 73, 70, 68, 75, 72, 
                              70, 72, 73, 70, 70, 72, 71, 70, 75, 72]})
#정규분포 구간 추정 함수 실행
lower, upper = stats.t.interval(0.95, len(df)-1, loc = np.mean(df), scale = scipy.stats.sem(df))
print("신뢰구간 : ({},{})".format(lower.round(2), upper.round(2)))

신뢰구간 : ([70.6],[72.4])


## 실습 12 p107

### 다음 상황에 대하여 모수의 신뢰구간을 계산한다.
 - 날아라 항공사의 콜센터 성능을 평가하기 위해 50번의 전화를 랜더만 시각에 걸어 상담원과 상담하기까지의 통화 대기 시간을 기록하였다. 고객들은 통화대기시간이 15초를 넘길 경우 짜증을 내기 시작하므로 15초를 초과한 통화대기는 통화 관리 실패로 규정하고있다.
 - 통화대기시간의 모분산에 대한 점추정량 및 95% 신뢰구간을 구하시오

In [14]:
df = pd.DataFrame({'sample':[12,9,10,5,16,13,5,9,8,13,
                            16,12,12,4,14,9,11,14,9,17,
                            5,8,9,6,11,19,6,6,8,9,
                            13,10,8,13,19,8,12,14,16,16,
                            11,17,5,9,12,17,11,8,5,8]})
df.describe()

sample
count  50.000000
mean   10.740000
std     3.983691
min     4.000000
25%     8.000000
50%    10.500000
75%    13.000000
max    19.000000

In [15]:
n = 50 #자료의 수 
s = 3.984 #위의 describe()를 통해 얻은 표준 편차
chi2 = 71.42 #chi^2(0.025,49)
chi2_1sub_alpha = 32.36 #chi^2(0.975,49)

trust_min=(n-1)*s*s/chi2 #신뢰구간 최소
trust_max =(n-1)*s*s/chi2_1sub_alpha #신뢰구간 최대
print("모분산의 95%신뢰구간 = ({:.3f},{:.3f})".format(trust_min,trust_max))

모분산의 95%신뢰구간 = (10.890,24.034)


## 실습 13 p110

### 다음 상황에 대하여 모수의 신뢰구간을 계산한다.
 - 날아라 항공사의 콜센터 성능을 평가하기 위해 50번의 전화를 랜더만 시각에 걸어 상담원과 상담하기까지의 통화 대기 시간을 기록하였다. 고객들은 통화대기시간이 15초를 넘길 경우 짜증을 내기 시작하므로 15초를 초과한 통화대기는 통화 관리 실패로 규정하고있다.
 - 통화관리 성공율에 대한 점추정량 및 95% 신뢰구간을 구하시오


In [16]:
import math

p = 9/50
n = 50
Z = 1.96

trust_min = p - Z*math.sqrt(p*(1-p)/n)
trust_max = p + Z*math.sqrt(p*(1-p)/n)

print("점추적량 = {}".format(p))
print("통화관리 성공율에 대한 95% 신뢰".format(trust_min, trust_max))

점추적량 = 0.18
통화관리 성공율에 대한 95% 신뢰


## 실습 14 p131

### 고객서비스센터의 고객만족도 평균은 78.0이 이다. 개선활동을 완료한 후 다음과 같은 10개의 고객만족도 데이터를 얻었다. 
- 개선활동이 만족도를 변화 시켰는가? ($\alpha =0.05$) 

In [17]:
df = pd.DataFrame({'sample':[85.0, 79.0, 79.1, 79.9, 81.6,
                             78.6, 85.4, 83.4, 78.1, 79.2]})
t_result = stats.ttest_1samp(df, 78.0)

t,p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("1-Sample t-test")
print("t 통계량 : {}".format(t))
print("p-value : {}".format(p))

1-Sample t-test
t 통계량 : [3.379]
p-value : [0.008]


$\alpha=0.05$ 수준에서 귀무가설을 기각하기 위해서 요구되는 최소한의 t값(임계값)은 $t_{(0.025,9)} =2.262$ 이기에 위의 경우 t의 값이 3.379 이므로 2.262보다 크므로 개선활동이 고객만족도 변화를 시켰다고 할 수 있다.

## 실습 15 p139

### 제품을 A공급자와 B공급자로 부터 제품을 받아 두께를 측정해서 다음의 데이터를 얻었다.
- 공급자 간에 모평균에 차이가 있다고 할 수 있는가? (유의수준 $ \alpha = 0.05$)

In [18]:
df = pd.DataFrame({'Thickness': [1.435, 1.572, 1.486, 1.883, 1.715, 1.799, 1.511, 1.457,
                              1.548, 1.768, 1.711, 1.832, 1.427, 1.344, 1.404, 1.883],
                  'Supplier' : ['B', 'B', 'B', 'A', 'A', 'A', 'B','B',
                                'B', 'A', 'A', 'A', 'A', 'A', 'B','B']})

df_A = df[df.Supplier =='A']
df_B = df[df.Supplier =='B']
df_A = df_A['Thickness']
df_B = df_B['Thickness']

t_result = stats.ttest_ind(df_A,df_B)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("2-Sample t-test")
print("t: {}".format(t))
print("p: {}".format(p))


2-Sample t-test
t: 1.7
p: 0.111


유의수준 $ \alpha = 0.05$ 이고 차이가 있는지를 알아보는것이기에 양측검정이다. 따라서 p value 와 비교해야할 대상은 0.05/2 = 0.025 이다.p값이 0.025보다 크기에 귀무가설을 채택한다. 따라서 두 공급자 간에 모평균의 차이가 없다고 할 수 있다.

## 실습 16 p143

### 회사원 10명에 대해 영어점수가 학원가기 전의 점수와 학원수강 후에 영어점수가 짝을 이루고 있다.
 - 회사원들의 성적은 학원가기 전후에 대해 차이가 있는지 검정하라.

In [19]:
df1 = pd.DataFrame({'before':[720, 589, 780, 648, 720, 589, 780, 648, 780, 648]})
df2 = pd.DataFrame({'after': [710, 580, 787, 712, 750, 600, 782, 670, 790, 680]})

t_result = stats.ttest_rel(df1,df2)

t,p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("Paired t-test")
print("t: {}".format(t))
print("p: {}".format(p))

Paired t-test
t: [-2.266]
p: [0.05]


유의수준이 정해져 있지 않으므로 유의수준($ \alpha = 0.01$) 로 정하겠다. 차이가 있는지 검정해야 하므로 $ \alpha /2= 0.005$ 와 p-value를 비교해야한다. 이때 p-value가 $ \alpha /2 = 0.005$ 보다 크므로 귀무가설을 채택한다. 따라서 회사원들의 성적은 학원가기 전후에 차이가 없다고 검정할 수 있다.